bazel/cas/client.go

```diff
@@ -277,3 +319,155 @@ func updateCacheFromClient(acc remoteexecution.ActionCacheClient,
 
 	return ar, nil
 }
+
+type BatchUploadContent struct {
+	Digest *remoteexecution.Digest
+	Data   []byte
+}
+
+/*
+client for uploading a list of digests to CAS BatchUpdateBlobs (as a single batch)
+ */
+func (casCli *CASClient)BatchUpdateWrite(r dialer.Resolver, contents []BatchUploadContent, b backoff.BackOff) ([]*remoteexecution.Digest, error) {
+
+	// make the request object
+	requests := make([]*remoteexecution.BatchUpdateBlobsRequest_Request, 0)
+	for _, content := range contents {
+		requests = append(requests, &remoteexecution.BatchUpdateBlobsRequest_Request{
+			Digest:               content.Digest,
+			Data:                 content.Data,
+		})
+	}
+	request := &remoteexecution.BatchUpdateBlobsRequest{
+		InstanceName:         bazel.DefaultInstanceName,
+		Requests:             requests,
+		XXX_NoUnkeyedLiteral: struct{}{},
+		XXX_unrecognized:     nil,
+		XXX_sizecache:        0,
+	}
+
+	var ctx context.Context = context.Background()
+
+	// upload with retries
+	try := 1
+	e := backoff.Retry(func() error {
+		var err error
+		log.Debugf("Try #%d", try)
+		casCli.uploadedDigests, err = casCli.batchUpdateWriter(r, request, ctx)
+		try += 1
+		return err
+	}, b)
+
+	// if errors are found return an error result
+	if e != nil {
+		return casCli.uploadedDigests, fmt.Errorf("CAS BatchUpdateBlobs() returned error:%s", e.Error())
+	}
+
+	return casCli.uploadedDigests, nil
```

Comment:
Why is uploadedDigests part of the CASClient? Looks like we can remove and track only in this function.

Reply:
done